File: 131.palindrome-partitioning.cpp
/*
 * @lc app=leetcode id=131 lang=cpp
 *
 * [131] Palindrome Partitioning
 */
#include <vector>
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>

using namespace std;

// @lc code=start
class Solution {
public:
    vector<vector<string>> partition(string s) {
        int n = s.length();

        vector<vector<string>> result;
        vector<string> curr_partition;
        // Backtracking algorithm
        backtrack(s, 0, 1, curr_partition, result);
        return result;
    }

    bool checkPalindrome(string& substring) {
        int i = 0;
        int j = substring.length() - 1;
        while (i<substring.length() && j >= 0 && i<=j) {
            if (substring[i] != substring[j]) {
                return false;
            }
            i++;
            j--;
        }
        return true;
    }

    void backtrack(string& s, int beg, int end, vector<string>& curr_partition, vector<vector<string>>& result) {
        if (beg == s.size()-1) {
            result.push_back(curr_partition);
        }

        if (end >= s.size())
            return;

        backtrack(s, beg, end+1, curr_partition, result);
        string substring = s.substr(beg, end-beg+1);
        int substring_length = substring.size();

        if (checkPalindrome(substring)) {
            curr_partition.push_back(substring);
            backtrack(s, beg+substring_length, beg+substring_length+1, curr_partition, result);
        }
    }
};
// @lc code=end

File: 131.palindrome-partitioning.py
#
# @lc app=leetcode id=131 lang=python3
#
# [131] Palindrome Partitioning
#
from typing import List, Dict, Set, Tuple

# @lc code=start
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        hm: Dict[str: List[str]] = {}

        def isPalindrome(s: str) -> bool:
            for i in range(len(s)//2):
                if s[i] != s[len(s)-i-1]:
                    return False
            return True

        def allPartitions(s: str) -> List[List[str]]:
            if hm.get(s, None) is not None:
                return hm[s]
            if len(s) == 1:
                hm[s] = [[s]]
                return hm[s]
            temp: Set[Tuple[str]] = set()
            if isPalindrome(s):
                temp.add((s,))
            i = 0
            for i in range(1, len(s)):
                # Partition the string into s[0:i-1] and s[i:]
                leftPartitions = allPartitions(s[0:i])
                rightPartitions = allPartitions(s[i:])
                for l in leftPartitions:
                    for r in rightPartitions:
                        temp.add(tuple(l + r))
            temp = [[string for string in partition] for partition in temp]
            hm[s] = temp
            return hm[s]
        
        return allPartitions(s)
# @lc code=end

def main():
    s = Solution()
    print(s.partition("aab"))

if __name__ == '__main__':
    main()